### ISSUE TITLE
Script loading performance

### ISSUE THREAD

**Issue by exectails** [aura-project/aura#300]
Currently we compile and load every single script file separately. This allows us to cache the individual scripts and on reload/restart only recompile those that changed. A good idea in theory, but in reality it's much, *much* faster to compile them all at once, into one assembly, if more than a handful of scripts have to be compiled.

In a quick test I was able to compile 5000 small test scripts in 5 seconds. Loading that many scripts with our current method would take hours, if not days. One has to keep in mind, right now we might only have 400 scripts, but we're barely done with *G1*.

Compiling everything at once would reduce the initial, uncached script loading *significantly*. Loading the one cached assembly might be a little faster, but I'm not sure how much. Whether reloading is faster or not will depend on how many scripts changed and where they are in the loading order. For example, a script that is at the end of the loading order can take well over 5 seconds to load, while the very first will compile in an instant. Loading everything will always take the same time though, as everything has to be compiled again.

In short: the script loading speed will improve in some cases, and in some it will stay approximately the same.

**Comment by exectails** [comment]
I forgot one more thing, which is either an advantage or a disadvantage, depending on how you look at it. If all files are compiled at once, we can't use a pre-compiler anymore, unless we create a temporary copy of every file. That means this couldn't be done before the AI rewrite, because our current AIs depend on the pre-compiler, and we'd have to start putting usings in the scripts.

```csharp
using Aura.Channel.Scripting.Scripts;
using Aura.Channel.World.Entities;
using Aura.Mabi;
using Aura.Mabi.Const;
using System.Threading.Tasks;

public class DuncanBaseScript : NpcScript
```

However, if we do that, we could create a project inside the solution for the scripts, and get proper Intellisense.

**Comment by exectails** [comment]
An alternative could be fixing the inclusion of *all* assemblies loaded before, and instead create a reference system. The scripts are supposed to be able to reference each other, but they don't need a reference to *all* others. For example, if you need Duncan for some reason, you might put this in your script:

```csharp
// #reference "npcs/tir/duncan.cs"
```

and as a result that one script would get a reference to Duncan's assembly. This doesn't speed up compilation as much as compiling it all at once, but it does speed it up significantly.

**Comment by QewQew** [comment]
Could you upload your test project for me to take a look at? Thanks

**Comment by exectails** [comment]
Sure, here you go. (GitHub didn't accept the ZIP for some reason.)
http://puu.sh/qcLyd/66e3fcba27.zip

**Comment by exectails** [comment]
Just realized there is one more thing that would change with this, one error in one script would stop the compilation of *everything*. Which means that if one script has an error, you suddenly don't have any NPCs, AIs, dungeons, etc anymore, until you fix the error.

I feel like we should rather try a reference approach, as mentioned above, which would give us performance *and* usability and flexibility. The user folder system is a hurdle in that, but I'm sure we could come up with a solution. For ease of use the scripts could also automatically reference all scripts in the same folder, which would be the most common case where you need references.

**Comment by xCryptic** [comment]
Perhaps make it an option of which compiler the user wants? That way they decide if they'd rather put up with long loading times, or a scriptless aura if there's an error.

**Comment by eotetb** [comment]
I like that idea, do you think that it is feasible to have two separate compilers?

**Comment by YaibaToKen** [comment]
If it's something that causes the server owner to rebuild the files whenever changing the used compiler then I assume it won't cause much bloat on the server files. it would also bring about a new layer of choice. On the other hand, I don't know how easy to maintain the two compilers at the same time would be

**Comment by exectails** [comment]
If our compiler interface is changed a little this actually wouldn't be all that complicated.

**Comment by exectails** [comment]
One more thing to note is that after this change we won't be able to script in VB and our beloved Boo anymore =PP

**Comment by xCryptic** [comment]
noooooooooooooooooooo my vb </3






ah well, was nice while it lasted (not that I really used it much)
